internal/factory/container/mounts.go

```diff
@@ -0,0 +1,285 @@
+
+// Examples:
+
+// isSubDirectoryOf("/var/lib/containers/storage", "/") returns true
+// isSubDirectoryOf("/var/lib/containers/storage", "/var/lib") returns true
+// isSubDirectoryOf("/var/lib/containers/storage", "/var/lib/containers") returns true
+// isSubDirectoryOf("/var/lib/containers/storage", "/var/lib/containers/storage") returns true
+// isSubDirectoryOf("/var/lib/containers/storage", "/var/lib/containers/storage/extra") returns false
+// isSubDirectoryOf("/var/lib/containers/storage", "/va") returns false
+// isSubDirectoryOf("/var/lib/containers/storage", "/var/tmp/containers") returns false
+func isSubDirectoryOf(base, target string) bool {
+	if !strings.HasSuffix(target, "/") {
+		target += "/"
+	}
+	if !strings.HasSuffix(base, "/") {
+		base += "/"
+	}
+	return strings.HasPrefix(base, target)
+}
+
+// resolveSymbolicLink resolves a possible symlink path. If the path is a symlink, returns resolved
+// path; if not, returns the original path.
+// note: strictly SecureJoin is not sufficient, as it does not error when a part of the path doesn't exist
+// but simply moves on. If the last part of the path doesn't exist, it may need to be created.
+func resolveSymbolicLink(scope, path string) (string, error) {
+	info, err := os.Lstat(path)
+	if err != nil {
+		return "", err
+	}
+	if info.Mode()&os.ModeSymlink != os.ModeSymlink {
+		return path, nil
+	}
+	if scope == "" {
+		scope = "/"
+	}
+	return securejoin.SecureJoin(scope, path)
+}
+
+func (c *container) addCriMount(mount *rspec.Mount) {
+	if mount != nil {
+		dst := filepath.Clean(mount.Destination)
+		if dst == "/dev" {
+			c.mountInfo.criDevSet = true
+		}
+		if dst == "/sys" {
+			c.mountInfo.criSysSet = true
+		}
+		c.mountInfo.criMounts[dst] = mount
+	}
+}
+
+func (c *container) isInCRIMounts(dst string) bool {
+	dst = filepath.Clean(dst)
+	if _, ok := c.mountInfo.criMounts[dst]; ok ||
+		(strings.HasPrefix(dst, "/dev/") && c.mountInfo.criDevSet) ||
+		(strings.HasPrefix(dst, "/sys/") && c.mountInfo.criSysSet) {
+		return true
+	}
+	return false
+}
+
+func (c *container) isInMounts(dst string) bool {
+	_, ok := c.mountInfo.mounts[dst]
+	return ok
+}
+
+func (c *container) addMount(mount *rspec.Mount) {
+	if mount != nil {
+		c.mountInfo.mounts[filepath.Clean(mount.Destination)] = mount
+	}
+}
+
+// specAddMounts add the mounts to the OCI Spec
+func specAddMounts(c *container) {
+	allmounts := make([]*rspec.Mount, 0, len(c.mountInfo.mounts))
+	for k := range c.mountInfo.mounts {
+		allmounts = append(allmounts, c.mountInfo.mounts[k])
+	}
+
+	// Sort & Add mounts to specgen
+	sort.Sort(orderedMounts(allmounts))
+	for _, m := range allmounts {
+		c.spec.AddMount(*m)
+	}
+}
+
+func newMountInfo() *mountInfo {
+	return &mountInfo{
+		criMounts: make(map[string]*rspec.Mount),
+		mounts:    make(map[string]*rspec.Mount),
+	}
+}
+
+func clearMountInfo(c *container) {
+	c.mountInfo.criMounts = nil
+	c.mountInfo.mounts = nil
+	c.mountInfo = nil
+}
```

Comment:
why not leave it for the duration of the container creation process? the container factory won't live that long